Deferring ThingHandler.initialize() until all XMLs have been parsed

sjka opened the pull request:
This would be a short-term solution as discussed in #1143. 

Long-term, it would still make sense to have a fast-loading index, e.g. generated at build-time in order to postpone loading of the XMLs until they're needed.

fixes #1143
Signed-off-by: Simon Kaufmann <simon.kfm@googlemail.com>

kaikreuzer commented:
@cdjackson Could you test if this solves your issue?

cdjackson commented:
Will do, but not till Friday as I’m travelling until then and don’t have ready access to ZWave dev environment…

> On 26 Apr 2016, at 12:48, Kai Kreuzer <notifications@github.com> wrote:
> 
> @cdjackson <https://github.com/cdjackson> Could you test if this solves your issue?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub <https://github.com/eclipse/smarthome/pull/1419#issuecomment-214700109>

cdjackson commented:
So, I've pulled in this branch, and enabled the local config.xml and core.thing bundles which looks like it's all that's changed in this PR (??). I can see that this is running as the asyncLoaderRegistration is being set, but I set a breakpoint on line 212 of AbstractAsyncBundleProcessor (which looks like it's where the check is made?) and this isn't being called.

Actually, it looks like this method is never called from anywhere, so maybe I'm not looking in the right place to see what's happening...

Bottom line though is it's not working -:


```
2016-05-01 11:27:09.513 [DEBUG] [c.x.o.XmlDocumentBundleTracker:125  ] - Reading the XML document '/ESH-INF/thing/honeywell_th8320zw_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-05-01 11:27:09.515 [DEBUG] [c.x.o.XmlDocumentBundleTracker:125  ] - Reading the XML document '/ESH-INF/thing/horstmann_asrzw_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-05-01 11:27:09.516 [DEBUG] [c.x.o.XmlDocumentBundleTracker:125  ] - Reading the XML document '/ESH-INF/thing/horstmann_hrt4zw_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-05-01 11:27:09.518 [DEBUG] [c.x.o.XmlDocumentBundleTracker:125  ] - Reading the XML document '/ESH-INF/thing/horstmann_scsc17_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-05-01 11:27:09.520 [DEBUG] [.core.common.ThreadPoolManager:141  ] - Created scheduled thread pool 'thingHandler' of size 5
2016-05-01 11:27:09.524 [DEBUG] [c.x.o.XmlDocumentBundleTracker:125  ] - Reading the XML document '/ESH-INF/thing/horstmann_sir321_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-05-01 11:27:09.525 [DEBUG] [.c.thing.internal.ThingManager:113  ] - Thing handler for thing 'zwave:serial_zstick:154680319e2' added.
2016-05-01 11:27:09.526 [DEBUG] [.c.thing.internal.ThingManager:333  ] - Assigning handler for thing 'zwave:serial_zstick:154680319e2'.
2016-05-01 11:27:09.527 [DEBUG] [c.x.o.XmlDocumentBundleTracker:125  ] - Reading the XML document '/ESH-INF/thing/horstmann_ssr302_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-05-01 11:27:09.530 [DEBUG] [c.x.o.XmlDocumentBundleTracker:125  ] - Reading the XML document '/ESH-INF/thing/intermatic_ha01c_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-05-01 11:27:09.531 [DEBUG] [.c.thing.internal.ThingManager:684  ] - All good. going to initialize.
2016-05-01 11:27:09.532 [DEBUG] [c.x.o.XmlDocumentBundleTracker:125  ] - Reading the XML document '/ESH-INF/thing/intermatic_ha04c_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-05-01 11:27:09.542 [DEBUG] [.c.thing.internal.ThingManager:688  ] - Calling initialize handler for thing 'zwave:serial_zstick:154680319e2' at 'org.openhab.binding.zwave.handler.ZWaveSerialHandler@31836b60'.
2016-05-01 11:27:09.543 [DEBUG] [c.x.o.XmlDocumentBundleTracker:125  ] - Reading the XML document '/ESH-INF/thing/kwikset_914trl_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-05-01 11:27:09.543 [DEBUG] [b.z.handler.ZWaveSerialHandler:67   ] - Initializing ZWave serial controller.
2016-05-01 11:27:09.544 [DEBUG] [o.b.z.h.ZWaveControllerHandler:76   ] - Initializing ZWave Controller.
2016-05-01 11:27:09.545 [INFO ] [b.z.handler.ZWaveSerialHandler:77   ] - Connecting to serial port '/dev/tty.SLAB_USBtoUART'
2016-05-01 11:27:09.553 [DEBUG] [c.x.o.XmlDocumentBundleTracker:125  ] - Reading the XML document '/ESH-INF/thing/leviton_dzmx11lz_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-05-01 11:27:09.554 [DEBUG] [c.x.o.XmlDocumentBundleTracker:125  ] - Reading the XML document '/ESH-INF/thing/leviton_dzpa1_0_0.xml' in module 'org.openhab.binding.zwave'...
2016-05-01 11:27:09.556 [DEBUG] [c.x.o.XmlDocumentBundleTracker:125  ] - Reading the XML document '/ESH-INF/thing/leviton_dzpd3_0_0.xml' in module 'org.openhab.binding.zwave'...
```

sjka commented:
Thanks @cdjackson for trying this out! The actual check is done at AbstractAsyncBundleProcessor:175, but nevertheless, the log output does not really look good. 
If I read your logs correctly, the Handler of the zwave:serial_zstick:... comes from the same bundle as the XML files that are being loaded - right? Therefore the serial stick's initialize() must not be called before all the XML files are loaded. 
I have no immediate idea what might have gone wrong, but I will try to figure out what it could be. If you have any ideas, please let me know.

cdjackson commented:
Thanks Simon - yes, the handler is in the same bundle as the XML files…

Let me know when you want me to try again :)

Cheers
Chris

sjka commented:
I have fixed a potential synchronization issue and added a whole bunch of debug logging to it, just in case this was not the issue in order to better understand what's going on. They will have to be removed before merge, of course...
@cdjackson could you please run this updated PR on your zwave environment again? Thanks

kaikreuzer commented:
@cdjackson Did you have a chance to test this?

cdjackson commented:
No - not yet. I’ll take a look tomorrow...

cdjackson commented:
Looks to be working fine - thanks.

kaikreuzer commented:
Thanks for testing!

kaikreuzer commented:
Hm, this PR seems to have broken some tests, see https://hudson.eclipse.org/smarthome/job/SmartHomeDistribution-Nightly/1039/testReport/
@sjka Could you please have a look?
